# **Coursera capstone project week 3 - part 1, 2 and 3**

####  by Francesca Scalesse

In this project we explore, segment, and cluster the neighborhoods in the city of Toronto based on the postalcode and borough information. The project will be divided in 3 part and every one will contains the previous ones. 

In this notebook it will be shown the all the three parts of the final assigment for the week 3 of the coursera course: *Applied data science capstone*.

## Part 1

First of all data about Toronto neighborhood will be scraped using BeautifulSoup. Then the acquired data will be transformed in a dataframe using pandas.

### Import needed libraries 

In [50]:
#!pip install pandas
#!pip install requests
!pip install bs4
#!pip install plotly

In [51]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

### Create a beautiful soup object

In the following section the data about the city of Toronto will be retrieved by the URL and a beautiful soup object will be created. The source of the data is the wikipedia webpage about Toronto Neighborhood data (https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M)

In [52]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [53]:
html_data = requests.get(url).text

In [54]:
soup = BeautifulSoup(html_data, "html5lib")

Now the beautiful soup object will be used to create a dataframe using pandas. The dataframe (df) will have 3 columns: Postal_code, Borough and Neighborhood. If cells has a Borough value "Not assigned" they will be ignored.If a cell has a borough but a "Not assigned"  neighborhood, then the neighborhood will be the same as the borough.

In [55]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)


df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})



Now the first 5 rows of the dataframe will be shown

In [56]:
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [57]:
df.shape

(103, 3)

This data frame has 103 rows and 3 columns

## Part 2

In this section the dataframe will be updated adding the coordinate columns. In this way it will be possible explore Toronto's neighborhood using foursquare in the third part of the project. The data will be imported from their url and transformed in a dataframe. The neede libraries are already imported.

### Import data and transform in a dataframe

The name of the first column 

In [58]:
url_geospatial = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv"
df2 = pd.read_csv(url_geospatial)
df2.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [59]:
df2.shape

(103, 3)

This data frame has 103 rows and 3 columns

In [60]:
print(list(df2)[0])
print(list(df)[0])

Postal Code
PostalCode


### Merge the dataframes

Now we need to merge our two dataframes in a new one named "df3". Both of them have the same number of rows so we can easily use the method merge(). However we need first change the name of thi first column of dataframe "df2". In this way it will be possible to merge the dataframes using the column 'PostalCode'

**Rename column**

In [61]:
df2.rename(columns = {'Postal Code':'PostalCode'},inplace=True)
df2.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

**Merge the dataframes**

In [62]:
df3 = pd.merge(left = df, right = df2, how = 'outer', on = 'PostalCode' )
df3.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

## Part 3

In this section neighborhoods of Toronto will be explored and clustered. The cluster methods will be KNN.

### Import the needed libraries

In [63]:
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [64]:
import requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [65]:
import numpy as np

### Mapping Totonto's neighborhood

We use geopy to find the coordinates of Toronto 

In [66]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Now we create a map of Toronto Neighborhood

In [67]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df3['Latitude'], df3['Longitude'], df3['Borough'], df3['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Exploring the neighborhood

Now, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

**Define Foursquare credential and vesion**

In [68]:
CLIENT_ID = 'Y2E14U4GP44DD3R0E1MLC51OSZ33I0QUJ1RTVXFGM2Y4NZBB' 
CLIENT_SECRET = '3LJYB5GP2QSSJ04W5F2ZGWOSVPOIVRN0FUMQ1IQYORDZ2OD4' 
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: Y2E14U4GP44DD3R0E1MLC51OSZ33I0QUJ1RTVXFGM2Y4NZBB
CLIENT_SECRET:3LJYB5GP2QSSJ04W5F2ZGWOSVPOIVRN0FUMQ1IQYORDZ2OD4


### Explore allToronto's neighborhood

A function to explore all the neigborhood will be created. It will create the foursquare API request url, get the request and return the information about the venues.

In [69]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now apply the function on all the dataframe about Toronto's neighborhood (df3)

In [70]:
# type your answer here
toronto_venues = getNearbyVenues(names=df3['Neighborhood'], latitudes=df3['Latitude'], longitudes=df3['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

Information about the dataframe *toronto_venues*:

In [71]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1   Fast Food Restaurant  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

In [72]:
toronto_venues.shape

(2119, 7)

This dataframe has 2131 rows and 7 columns

Let's check how many venues were returned for each neighborhood

In [73]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               9   
Bathurst Manor, Wilson Heights, Downsview North                     22   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   26   
...                                                                ...   
Willowdale, Newtonbrook                                              1   
Woburn                                                               4   
Woodbine Heights                                                     7   
York Mills West                                                      2   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                9   
Bathurst Manor, Wilson Heights, Downsview North                      22   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    26   
...                                                                 ...   
Willowdale, Newtonbrook                                               1   
Woburn                                                                4   
Woodbine Heights                                                      7   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               9               9   
Bathurst Manor, Wilson Heights, Downsview North     22              22   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   26              26   
...                                                ...             ...   
Willowdale, Newtonbrook                              1               1   
Woburn                                               4               4   
Woodbine Heights                                     7               7   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         9   
Bathurst Manor, Wilson Heights, Downsview North               22   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             26   
...                                                          ...   
Willowdale, Newtonbrook                                        1   
Woburn                                                         4   
Woodbine Heights                                               7   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        9 

Let's find out how many unique categories can be curated from all the returned venues

In [74]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 271 uniques categories.


###  Analyzing each neighborhood

In [75]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  American Restaurant  \
0                   0               0                0                    0   
1                   0               0                0                    0   
2                   0               0                0                    0   
3                   0               0                0                    0   
4                   0               0                0                    0   

   Antique Shop  ...  Train Station  Vegetarian / Vegan Restaurant  \
0             0  ...              0                              0   
1             0  ...              0                              0   
2             0  ...              0                              0   
3             0  ...              0                              0   
4             0  ...              0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0         0          0            0              0  
1         0          0            0              0  
2         0          0            0              0  
3         0          0            0              0  
4         0          0            0              0  

[5 rows x 271 columns]

In [76]:
toronto_onehot.shape

(2119, 271)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [77]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                0.0             0.0                0.0      0.0   
1                0.0             0.0                0.0      0.0   
2                0.0             0.0                0.0      0.0   
3                0.0             0.0                0.0      0.0   
4                0.0             0.0                0.0      0.0   

   Airport Food Court  Airport Lounge  Airport Service  American Restaurant  \
0                 0.0             0.0              0.0             0.000000   
1                 0.0             0.0              0.0             0.000000   
2                 0.0             0.0              0.0             0.000000   
3                 0.0             0.0              0.0             0.000000   
4                 0.0             0.0              0.0             0.038462   

   ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...            0.0                            0.0               0.0   
1  ...            0.0                            0.0               0.0   
2  ...            0.0                            0.0               0.0   
3  ...            0.0                            0.0               0.0   
4  ...            0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0          0.0                    0.0              0.0       0.0        0.0   
1          0.0                    0.0              0.0       0.0        0.0   
2          0.0                    0.0              0.0       0.0        0.0   
3          0.0                    0.0              0.0       0.0        0.0   
4          0.0                    0.0              0.0       0.0        0.0   

   Wings Joint  Women's Store  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0  

[5 rows x 271 columns]

In [78]:
toronto_grouped.shape

(101, 271)

Let's print each neighborhood along with the top 5 most common venues


In [79]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3         Chinese Restaurant  0.25
4                Yoga Studio  0.00


----Alderwood, Long Branch----
         venue  freq
0  Pizza Place  0.22
1     Pharmacy  0.11
2         Pool  0.11
3          Pub  0.11
4  Coffee Shop  0.11


----Bathurst Manor, Wilson Heights, Downsview North----
            venue  freq
0            Bank  0.09
1     Coffee Shop  0.09
2     Bridal Shop  0.05
3  Ice Cream Shop  0.05
4     Supermarket  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1  Japanese Restaurant  0.25
2   Chinese Restaurant  0.25
3                 Bank  0.25
4          Yoga Studio  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0  Italian Restaurant  0.08
1      Sandwich Place  0.08
2         Coffee Shop  0.08
3          Restaurant  0.08
4                 Pub  0.04

**Let's put that into a pandas dataframe**

First, let's write a function to sort the venues in descending order.


In [80]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [81]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant    Chinese Restaurant   
1              Pharmacy                       Pool                   Pub   
2           Coffee Shop                Bridal Shop        Ice Cream Shop   
3   Japanese Restaurant         Chinese Restaurant                  Bank   
4        Sandwich Place                Coffee Shop            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio    Miscellaneous Shop   Monument / Landmark   
1           Coffee Shop          Skating Rink                   Gym   
2           Supermarket   Fried Chicken Joint        Sandwich Place   
3           Yoga Studio                 Motel   Monument / Landmark   
4                   Pub               Butcher                  Café   

             8th Most Common Venue       9th Most Common Venue  \
0  Molecular Gastronomy Restaurant  Modern European Restaurant   
1                   Sandwich Place    Mediterranean Restaurant   
2                      Gas Station                  Restaurant   
3  Molecular Gastronomy Restaurant  Modern European Restaurant   
4                   Breakfast Spot                    Boutique   

  10th Most Common Venue  
0      Mobile Phone Shop  
1            Men's Store  
2      Mobile Phone Shop  
3      Mobile Phone Shop  
4           Liquor Store

### Clustering with KNN

Run k-means to cluster the neighborhood into 5 clusters.


In [82]:
from sklearn.cluster import KMeans

In [83]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 4, 4, 4, 4, 4, 4, 4, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [84]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df3

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             0.0  Fast Food Restaurant                    Park   
1 -79.315572             4.0          Intersection             Pizza Place   
2 -79.360636             4.0           Coffee Shop                  Bakery   
3 -79.464763             4.0        Clothing Store  Furniture / Home Store   
4 -79.389494             4.0           Coffee Shop        Sushi Restaurant   

        3rd Most Common Venue  4th Most Common Venue  \
0           Food & Drink Shop          Metro Station   
1  Financial or Legal Service  Portuguese Restaurant   
2                        Park                Theater   
3                    Boutique            Coffee Shop   
4                 Yoga Studio         Sandwich Place   

        5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0  Modern European Restaurant      Mobile Phone Shop    Miscellaneous Shop   
1                Hockey Arena            Coffee Shop    Miscellaneous Shop   
2              Breakfast Spot                    Pub                  Café   
3           Accessories Store  Vietnamese Restaurant    Miscellaneous Shop   
4                    Beer Bar    Fried Chicken Joint         Burrito Place   

       8th Most Common Venue            9th Most Common Venue  \
0  Middle Eastern Restaurant               Mexican Restaurant   
1        Moroccan Restaurant              Monument / Landmark   
2                 Shoe Store                       Beer Store   
3        Arts & Crafts Store  Molecular Gastronomy Restaurant   
4              Smoothie Shop               Mexican Restaurant   

            10th Most Common Venue  
0         Mediterranean Restaurant  
1  Molecular Gastronomy Restaurant  
2                French Restaurant  
3       Modern European Restaurant  
4                             Café

Import folium map rendering library. Finally, let's visualize the resulting clusters. However, considering there are some NaN value in the column "Cluster Labels" we need to change. We converted to the integer 0.

In [85]:
toronto_merged['Cluster Labels'].head(10)

0    0.0
1    4.0
2    4.0
3    4.0
4    4.0
5    NaN
6    2.0
7    4.0
8    4.0
9    4.0
Name: Cluster Labels, dtype: float64

Replace NaN value with 0

In [86]:
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].replace(np.nan, 0)
toronto_merged['Cluster Labels']. head(10)

0    0.0
1    4.0
2    4.0
3    4.0
4    4.0
5    0.0
6    2.0
7    4.0
8    4.0
9    4.0
Name: Cluster Labels, dtype: float64

Transform data type  from float to integer

In [87]:
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)
toronto_merged['Cluster Labels'].head()

0    0
1    4
2    4
3    4
4    4
Name: Cluster Labels, dtype: int64

Display the map of the clusters

In [ ]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine the clusters

**Cluster 1**

In [65]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             0.0  Fast Food Restaurant   
21                    York             0.0                  Park   
35  East York/East Toronto             0.0                  Park   
40              North York             0.0            Playground   
52              North York             0.0                  Park   
61         Central Toronto             0.0                  Park   
64                    York             0.0                  Park   
66              North York             0.0                  Park   
83         Central Toronto             0.0                  Park   
85             Scarborough             0.0            Playground   
91        Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue  \
0                   Park     Food & Drink Shop   
21         Women's Store                  Pool   
35     Convenience Store         Metro Station   
40               Airport                  Park   
52           Yoga Studio         Metro Station   
61              Bus Line           Swim School   
64     Convenience Store           Yoga Studio   
66     Convenience Store           Yoga Studio   
83            Restaurant           Yoga Studio   
85          Intersection                  Park   
91            Playground                 Trail   

              4th Most Common Venue            5th Most Common Venue  \
0                     Metro Station       Modern European Restaurant   
21                    Metro Station  Molecular Gastronomy Restaurant   
35                      Yoga Studio               Mexican Restaurant   
40               Mexican Restaurant              Monument / Landmark   
52  Molecular Gastronomy Restaurant       Modern European Restaurant   
61                      Yoga Studio               Mexican Restaurant   
64               Mexican Restaurant  Molecular Gastronomy Restaurant   
66               Mexican Restaurant  Molecular Gastronomy Restaurant   
83               Mexican Restaurant  Molecular Gastronomy Restaurant   
85                      Yoga Studio               Mexican Restaurant   
91                      Yoga Studio                    Metro Station   

              6th Most Common Venue       7th Most Common Venue  \
0                 Mobile Phone Shop          Miscellaneous Shop   
21       Modern European Restaurant           Mobile Phone Shop   
35  Molecular Gastronomy Restaurant  Modern European Restaurant   
40  Molecular Gastronomy Restaurant  Modern European Restaurant   
52                Mobile Phone Shop          Miscellaneous Shop   
61       Modern European Restaurant           Mobile Phone Shop   
64       Modern European Restaurant           Mobile Phone Shop   
66       Modern European Restaurant           Mobile Phone Shop   
83       Modern European Restaurant           Mobile Phone Shop   
85  Molecular Gastronomy Restaurant  Modern European Restaurant   
91       Modern European Restaurant           Mobile Phone Shop   

        8th Most Common Venue      9th Most Common Venue  \
0   Middle Eastern Restaurant         Mexican Restaurant   
21         Miscellaneous Shop  Middle Eastern Restaurant   
35          Mobile Phone Shop         Miscellaneous Shop   
40          Mobile Phone Shop         Miscellaneous Shop   
52  Middle Eastern Restaurant         Mexican Restaurant   
61         Miscellaneous Shop  Middle Eastern Restaurant   
64         Miscellaneous Shop  Middle Eastern Restaurant   
66         Miscellaneous Shop  Middle Eastern Restaurant   
83         Miscellaneous Shop  Middle Eastern Restaurant   
85          Mobile Phone Shop         Miscellaneous Shop   
91         Miscellaneous Shop  Middle Eastern Restaurant   

       10th Most Common Venue  
0    Mediterranean Restaurant  
21         Mexican Restaurant  
35  Middle Eastern Restaurant  
40  Middle Eastern Restaurant  
52                Men's Store  
61              Me

**Cluster 2**

In [60]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
53   North York             1.0            Food Truck        Baseball Field   
57   North York             1.0        Baseball Field           Yoga Studio   
101   Etobicoke             1.0        Baseball Field           Yoga Studio   

         3rd Most Common Venue       4th Most Common Venue  \
53               Metro Station  Modern European Restaurant   
57   Middle Eastern Restaurant         Monument / Landmark   
101  Middle Eastern Restaurant         Monument / Landmark   

               5th Most Common Venue       6th Most Common Venue  \
53                 Mobile Phone Shop          Miscellaneous Shop   
57   Molecular Gastronomy Restaurant  Modern European Restaurant   
101  Molecular Gastronomy Restaurant  Modern European Restaurant   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
53   Middle Eastern Restaurant    Mexican Restaurant           Yoga Studio   
57           Mobile Phone Shop    Miscellaneous Shop    Mexican Restaurant   
101          Mobile Phone Shop    Miscellaneous Shop    Mexican Restaurant   

    10th Most Common Venue  
53     Monument / Landmark  
57           Luggage Store  
101          Luggage Store

**Cluster 3**

In [61]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant   Moroccan Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6                Market   Martial Arts School        Massage Studio   

  6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
6        Medical Center  Mediterranean Restaurant           Men's Store   

  9th Most Common Venue 10th Most Common Venue  
6         Metro Station     Mexican Restaurant

**Cluster 4**

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11    Etobicoke             3.0                Bakery           Yoga Studio   
32  Scarborough             3.0            Playground         Grocery Store   

        3rd Most Common Venue 4th Most Common Venue  \
11  Middle Eastern Restaurant   Monument / Landmark   
32                Yoga Studio    Mexican Restaurant   

              5th Most Common Venue            6th Most Common Venue  \
11  Molecular Gastronomy Restaurant       Modern European Restaurant   
32              Monument / Landmark  Molecular Gastronomy Restaurant   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
11           Mobile Phone Shop    Miscellaneous Shop    Mexican Restaurant   
32  Modern European Restaurant     Mobile Phone Shop    Miscellaneous Shop   

       10th Most Common Venue  
11              Luggage Store  
32  Middle Eastern Restaurant

**Cluster 5**

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1               North York             4.0          Intersection   
2         Downtown Toronto             4.0           Coffee Shop   
3               North York             4.0        Clothing Store   
4             Queen's Park             4.0           Coffee Shop   
7               North York             4.0  Caribbean Restaurant   
..                     ...             ...                   ...   
97        Downtown Toronto             4.0           Coffee Shop   
98               Etobicoke             4.0                 River   
99        Downtown Toronto             4.0           Coffee Shop   
100  East Toronto Business             4.0    Light Rail Station   
102              Etobicoke             4.0     Convenience Store   

      2nd Most Common Venue       3rd Most Common Venue  \
1               Pizza Place  Financial or Legal Service   
2                    Bakery                        Park   
3    Furniture / Home Store                    Boutique   
4          Sushi Restaurant                 Yoga Studio   
7       Japanese Restaurant                        Café   
..                      ...                         ...   
97                     Café                       Hotel   
98              Yoga Studio          Mexican Restaurant   
99         Sushi Restaurant         Japanese Restaurant   
100                     Spa               Burrito Place   
102           Grocery Store                 Social Club   

               4th Most Common Venue       5th Most Common Venue  \
1              Portuguese Restaurant                Hockey Arena   
2                            Theater              Breakfast Spot   
3                        Coffee Shop           Accessories Store   
4                     Sandwich Place                    Beer Bar   
7                                Gym              Baseball Field   
..                               ...                         ...   
97                        Restaurant                         Gym   
98   Molecular Gastronomy Restaurant  Modern European Restaurant   
99                           Gay Bar                  Restaurant   
100                       Restaurant                     Brewery   
102             Fast Food Restaurant              Sandwich Place   

     6th Most Common Venue     7th Most Common Venue  \
1              Coffee Shop        Miscellaneous Shop   
2                      Pub                      Café   
3    Vietnamese Restaurant        Miscellaneous Shop   
4      Fried Chicken Joint             Burrito Place   
7        Mobile Phone Shop       Monument / Landmark   
..                     ...                       ...   
97     Japanese Restaurant       American Restaurant   
98       Mobile Phone Shop        Miscellaneous Shop   
99             Yoga Studio  Mediterranean Restaurant   
100          Garden Center                Skate Park   
102        Supplement Shop                    Bakery   

               8th Most Common Venue            9th Most Common Venue  \
1                Moroccan Restaurant              Monument / Landmark   
2                         Shoe Store                       Beer Store   
3                Arts & Crafts Store  Molecular Gastronomy Restaurant   
4                      Smoothie Shop               Mexican Restaurant   
7    Molecular Gastronomy Restaurant       Modern European Restaurant   
..                               ...                              ...   
97                        Steakhouse                      Salad Place   
98         Middle Eastern Restaurant                    Metro Station   
99                       Men's Store             Fast Food Restaurant   
100                           Garden                       Smoke Shop   
102                   Discount Store           Thrift / Vintage Store   

              10th Most Common Venue  
1    Molecular Gastronomy Restaurant  
2                  French Restaurant  
3        